### PR TITLE
Fix CANBus crashing [WIP]

### PR DESCRIPTION
**Pull request by ClintLiddick**
These two changes seem to have resolved the problem at least on the IXXAT cards. I want to do more thorough testing, but so far we have > 3 hours of run-time without crashing, including some simple trajectories, for a single arm (left).

This accidentally also includes #5

### CODE REVIEW

File: include/barrett/bus/abstract/communications_bus.h [hunk 1]
```diff
@@ -45,7 +45,7 @@ namespace bus {
 class CommunicationsBus {
 public:
 	static const size_t MAX_MESSAGE_LEN = 8;  /** The maximum of any of the available communications buses */
-	static const double TIMEOUT = 1.0;  /** Bus connection timeout limit in seconds */
+	static const double TIMEOUT = 0.01;  /** Bus connection timeout limit in seconds */
```
Comment (mkoval) [inline include/barrett/bus/abstract/communications_bus.h:48]: Is this change necessary? It seems like the "connection timeout" only applies on startup, so it's fine to leave it relatively high.

File: src/os.cpp [hunk 2]
```diff
@@ -95,7 +95,8 @@ namespace barrett {
 
 void btsleep(double duration_s)
 {
-	assert(duration_s > 1e-6);  // Minimum duration is 1 us
+	// assert(duration_s > 1e-6);  // Minimum duration is 1 us
+	duration_s = std::max(duration_s, 1e-6);
```
Comment (mkoval) [inline src/os.cpp:99]: This change seems to be a duplicate of #5.

**Comment by mkoval** [comment]
Can you explain how this fixed the issues on the IXXAT card? It looks like the only change is to set `realtime` to `false`, instead of `true`, in two places. It's not obvious how this leads to a crash. 😁

**Comment by psigen** [comment]
The primary suspicion that led us to this was that `realtime` flags were being used to determine the code path for `btsleep()`  in certain places within socket handling code.

The issues that we saw seemed to be connected to thread starvation in calling `recv_from()` on the CAN socket fd, so we were investigating this relationship.

We don't exactly know the mechanism here, this is just a WIP to test whether the race condition is resolved (since we have to run for hours to diff whether crashes occur).

File: src/os.cpp (hunk 2 above)
Comment (ClintLiddick) [reply src/os.cpp:99]: Yup, in the PR comments.

File: include/barrett/bus/abstract/communications_bus.h (hunk 1 above)
Comment (ClintLiddick) [reply include/barrett/bus/abstract/communications_bus.h:48]: This is actually used every time when waiting for response can frames (it receives, sleeps, loops until total passed time is greater than this timeout whenever it's waiting for a response frame). I tested late last night and it seems like it's not necessary, although 1 second is absurd for this, because we're way past the ability to recover from a missed message. 10 ms is potentially recoverable in software (or at least can exit more gracefully).

**Comment by ClintLiddick** [comment]
Right, because of the hardcoded `realtime=true` flags we were actually skipping a lot of the update and buffering logic. I still feel like there is some unexplained behavior based on the syscall logging, especially the second thread picking up and receiving messages after the first thread exited in error, but it seems like properly enabling all the non-realtime logic avoids that error state in our case.

**Comment by ClintLiddick** [comment]
I also think I should replace these flags entirely with pre-processor macros, because most of the non-/realtime logic is wrapped in `#ifdef`s, and never accessible in runtime branches.

File: include/barrett/bus/abstract/communications_bus.h (hunk 1 above)
Comment (mkoval) [reply include/barrett/bus/abstract/communications_bus.h:48]: 10 ms makes sense. In that case, the comment "*bus connection timeout* limit in seconds" is misleading. Can you change that to be accurate?

Comment (psigen) [reply include/barrett/bus/abstract/communications_bus.h:48]: At some point, can we think a bit about whether we are minimally fixing this library with the intent of merging these changes upstream to make a more stable single repository, or just changing anything we think is wrong for our system?

It's possible this timeout has some purpose being long on another system configuration.  It's also not really related to the issue at hand.  Should we tie this to the other changes dealing with the `realtime` flag usage?
